edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.Service/Hosting.cs

```diff
@@ -31,8 +35,51 @@ public class Hosting
+                           This implies that the ClientCertificateRequired should be false.
+                           However, the naming of this field is not semantically correct.
+                           We looked at how aspnetcore used this field, and found that it is true when CertificateMode.AllowCertificate is set (which does not require certificates).
+                           Therefore we can set this field to true and it will still work with direct methods.
+                           */
+                        options.ClientCertificateRequired = true;
+                        options.RemoteCertificateValidationCallback = (_, clientCert, chain, policyErrors) =>
+                        {
+                            if (policyErrors != SslPolicyErrors.None)
+                            {
+                                return false;
+                            }
+
+                            Option<X509Chain> chainOption = chain == null ? Option.None<X509Chain>() : Option.Some(chain);
+                            TlsConnectionFeatureExtended tlsConnectionFeatureExtended = GetConnectionFeatureExtended(chainOption);
+                            context.Features.Set<ITlsConnectionFeatureExtended>(tlsConnectionFeatureExtended);
+                            return true;
+                        };
+                    }
+                },
+                /* ClientCertificateMode.AllowCertificate is the intuitive choice here, however is not what we need
+                   If we use this then aspnetcore will register it's own certificate validation callback when creating the SslStream.
+                   This aspnetcore-created certificate validation callback in turn calls the HttpsConnectionAdapterOptions.ClientCertificateValidation, which we can configure In HttpsConnectionAdapterOptions.
+                   This is not the desired solution for us because HttpsConnectionAdapterOptions.ClientCertificateValidation does not provide access to the client cert chain when the connection is in scope.
+
+                   Thus, we need to pass NoCertificate to block aspnet from creating the SslStream with the certificate validation callback.
+                   This will let us use HttpsConnectionAdapterOptions.OnAuthenticate in order to register certificate validation with the connection in scope.
+                   We can handle the cert / no-cert case inside of OnAuthenticate */
+                ClientCertificateMode = ClientCertificateMode.NoCertificate,
+                SslProtocols = sslProtocols
+            };
+
             int port = configuration.GetValue("httpSettings:port", 443);
-            var certificateMode = clientCertAuthEnabled ? ClientCertificateMode.AllowCertificate : ClientCertificateMode.NoCertificate;
             IWebHostBuilder webHostBuilder = new WebHostBuilder()
```

Comment:
Should we use HostBuilder instead of WebHostBuilder? https://docs.microsoft.com/en-us/aspnet/core/migration/22-to-30?view=aspnetcore-3.1&tabs=visual-studio#hostbuilder-replaces-webhostbuilder

Reply:
I spent some time trying to convert webhost to generic host, but they have made changes affecting autofac making a larger refactor necessary. I talked with Varun earlier and we decided to ignore it.

Reply:
I think we should handle it in another PR.